src/Database/Expression/CaseStatementExpression.php

```diff
@@ -102,48 +101,52 @@ class CaseStatementExpression implements CaseExpressionInterface
     /**
      * Constructor.
      *
-     * @param \Cake\Database\TypeMap|null $typeMap The type map to use when using an array of conditions for the `WHEN`
-     *  value.
+     * When a value is set, the syntax generated is
+     * `CASE case_value WHEN when_value ... END` (simple case),
+     * where the `when_value`'s are compared against the
+     * `case_value`.
+     *
+     * When no value is set, the syntax generated is
+     * `CASE WHEN when_conditions ... END` (searched case),
+     * where the conditions hold the comparisons.
+     *
+     * Note that `null` is a valid case value, and thus should
+     * only be passed if you actually want to create the simple
+     * case expression variant!
+     *
+     * @param \Cake\Database\ExpressionInterface|object|scalar|null $value The case value.
+     * @param string|null $valueType The case value type. If no type is provided, the type will be tried to be inferred
+     *  from the value.
      */
-    public function __construct(?TypeMap $typeMap = null)
+    public function __construct($value = null, ?string $valueType = null)
     {
-        if ($typeMap === null) {
-            $typeMap = new TypeMap();
-        }
-        $this->_typeMap = $typeMap;
-    }
+        if (func_num_args() > 0) {
+            if (
+                $value !== null &&
+                !is_scalar($value) &&
+                !(is_object($value) && !($value instanceof Closure))
```

Comment:
```suggestion
                !($value instanceof Closure)
```
This can be simplified to just the instance check, right?

Reply:
No, looks can be deceiving on this one, we want to accept objects but not closures, and since `is_object()` is `true` for closures, both checks are needed.

Reply:
aaah, right, sorry. I didn't see the logic until after I decided to ask and assumed it was what I thought.